### PR TITLE
Update NullQubit Allocate and Release functionalities

### PR DESCRIPTION
**Pull request by AmintorDusko**
**Context:** Running more complex workflows with Catalyzed NullQubit requires more structure than expected for the NullQubit class. 

**Description of the Change:** This PR will expand NullQubit's Allocate and Release Qubit functionalities.

[sc-76168]

### CODE REVIEW

**Comment by AmintorDusko** [comment]
> Do we need the qubit manager?

I don't know if we need it. I'm trying to mimic as closely as possible the LightningQubit behavior and this is very lightweight. 
I'm happy to go with any better suggestion. Just let me know. :slightly_smiling_face:

**Comment by erick-xanadu** [comment]
> I don't know if we need it. I'm trying to mimic as closely as possible the LightningQubit behavior and this is very lightweight.
I'm happy to go with any better suggestion. Just let me know. 🙂

Yeah, I think that's a good justification. It would mean that we would also profile the qubit manager when we profile lightning qubit which I think is the intention behind this story.

**Review by erick-xanadu — APPROVED** [review]

**Comment by dime10** [comment]
> Yeah, I think that's a good justification. It would mean that we would also profile the qubit manager when we profile lightning qubit which I think is the intention behind this story.

Hmm I'm not sure I agree, everything behind the QuantumDevice interface is a device-specific implementation detail, so the qubit manager we implemented for lightning is not guaranteed to appear in other devices 🤔 

A simpler implementation could just hand out `num_qubits_` as an ID on allocation, although you're right that the QubitManager is already lightweight to begin with so it probably doesn't make much of a difference.

**Review by dime10 — APPROVED** [review]
Thanks for providing this fix 💯 I'm happy to merge this although as mentioned in my other comment I think we can do without the qubit manager.

File: doc/releases/changelog-dev.md [hunk 1]
```diff
@@ -270,7 +273,10 @@
 
 <h3>Bug fixes</h3>
 
-* Resolve a bug where `mitigate_with_zne` does not work properly with shots and devices 
+* Resolve a bug where NullQubit was allocating all qubits to zero.
+  [(#1210)](https://github.com/PennyLaneAI/catalyst/pull/1210)
```
[the remark this answers is not in the document]
Comment (AmintorDusko) [reply doc/releases/changelog-dev.md:277]: OK. Thanks!

Comment (AmintorDusko) [inline doc/releases/changelog-dev.md:277]:
```suggestion
```

**Comment by AmintorDusko** [comment]
Thank you both for your approval. I agree that we could go without the Qubit Manager. 
As we agree that this is a lightweight solution, and this is blocking other work, I will merge it as is.

**Pull request merged**